File: bin/Data/AUI/Periodic/Sources/code_amenuitem.cpp
// AMenuItem and AMenuItemSource application source code
#include <TurboBadger/tb_widgets.h>
#include <Urho3D/AUI/AMenuWindow.h>
#include <Urho3D/AUI/AMenubar.h>
#include "PeriodicApp.h"

void PeriodicApp::setup_amenuitem( AWidget *layout )
{
    PODVector<AWidget*> dest;
    layout->SearchWidgetClass( "TBButton", dest );
    for (unsigned ii = 0; ii < dest.Size(); ii++)
        SubscribeToEvent(dest[ii], E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleAmenuitemEvent ));

    AMenuItemSource *mis = new AMenuItemSource(context_);
    mis->AddItem( new AMenuItem( context_,"ASelectItem1", "item1" ) );
    mis->AddItem( new AMenuItem( context_,"ASelectItem2", "item2", "Ctrl+." ) );
    mis->AddItem( new AMenuItem( context_,"ASelectItem3", "item3", "Ctrl+A", "DuckButton" ) );
    mis->AddItem( new AMenuItem( context_,"ASelectItem4", "item4", "Ctrl+O", "LogoAtomic" ) );
    context_->SetGlobalVar( "menuitemptr", mis );
}

void PeriodicApp::HandleAmenuitemEvent(StringHash eventType, VariantMap& eventData)
{
    using namespace WidgetEvent;
    AWidget* widget = static_cast<AWidget*>(eventData[P_TARGET].GetPtr());
    String refid = eventData[P_REFID].GetString();
    if ( widget == NULL ) return;
    if (eventData[P_TYPE] == UI_EVENT_TYPE_CLICK)
    {
        if (widget->GetId() == "uimenuitemcode" )
        {
            AppLog( "AMenuItem support : " + widget->GetId() + " was pressed " );
            ViewCode ( "Sources/code_amenuitem.cpp", widget->GetParent() );
        }
        if (widget->GetId() ==  "uimenuitemlayout" )
        {
            AppLog( "AMenuItem support : " + widget->GetId() + " was pressed ");
            ViewCode ( "Scenes/layout_amenuitem.ui.txt", widget->GetParent() );
        }

        if (widget->GetId() ==  "uimenuitempush" )
        {
            AMenuItemSource *gmis = (AMenuItemSource*)context_->GetGlobalVar("menuitemptr").GetPtr();
            AMenuWindow* mymenuwindow = new AMenuWindow( context_, widget, "MenuItemDemo");
            int xx = widget->GetX() + (widget->GetWidth()/2);
            int yy = widget->GetY() + (widget->GetHeight()/2);
            SubscribeToEvent(mymenuwindow, E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleAmenuitemEvent ));
            mymenuwindow->Show(gmis, xx, yy);
        }

        if (widget->GetId() ==  "MenuItemDemo" )
        {
            AppLog( "AMenuItem event : " + widget->GetId() + " and " + refid + " was selected ");
        }

        if (widget->GetId() ==  "uimi1" )
        {
            AppLog( "AMenuItem action : " + widget->GetId() + " was pressed ");
            AMenuItemSource *gmis = (AMenuItemSource*)context_->GetGlobalVar("menuitemptr").GetPtr();
            gmis->AddItem( new AMenuItem( context_, "New AMenuItem") );
        }
        if (widget->GetId() ==  "uimi2" )
        {
            AppLog( "AMenuItem action : " + widget->GetId() + " was pressed ");
            AMenuItemSource *gmis = (AMenuItemSource*)context_->GetGlobalVar("menuitemptr").GetPtr();
            gmis->AddItem( new AMenuItem( context_, "Newer AMenuItem", "neweritem" ) );
        }
        if (widget->GetId() ==  "uimi3" )
        {
            AppLog( "AMenuItem action : " + widget->GetId() + " was pressed ");
            AMenuItemSource *gmis = (AMenuItemSource*)context_->GetGlobalVar("menuitemptr").GetPtr();
            gmis->AddItem( new AMenuItem( context_, "A Duck", "aduck", "", "DuckButton" ) );
        }
        if (widget->GetId() ==  "uimi4" )
        {
            AppLog( "AMenuItem action : " + widget->GetId() + " was pressed ");
            AMenuItemSource *gmis = (AMenuItemSource*)context_->GetGlobalVar("menuitemptr").GetPtr();
            gmis->AddItem( new AMenuItem( context_, "Atomic!", "atomic", "", "LogoAtomic" ) );
        }
        if (widget->GetId() ==  "uimi5" )
        {
            AppLog( "AMenuItem action : " + widget->GetId() + " was pressed ");
            AMenuItemSource *gmis = (AMenuItemSource*)context_->GetGlobalVar("menuitemptr").GetPtr();
            gmis->Clear();
        }
    }
}

File: bin/Data/AUI/Periodic/LuaScripts/code_amenuitem.lua
-- AMenuItem and AMenuItemSource application source code

-- a global, persist this way, wouldnt do it thru a SetGlobalVar
mgmis = nil

function setup_uimenuitem( layout )
    local dest = layout:SearchAWidgetClass( "TBButton" )
    for i, awid in ipairs(dest) do
        SubscribeToEvent(awid, "WidgetEvent", "HandleUimenuitemEvent")
	end

    mis = AMenuItemSource:new()
    mis:AddItem( AMenuItem:new( "ASelectItem1", "item1" ) )
    mis:AddItem( AMenuItem:new( "ASelectItem2", "item2", "Ctrl+X" ) )
    mis:AddItem( AMenuItem:new( "ASelectItem3", "item3", "Ctrl+D", "DuckButton" ) )
    mis:AddItem( AMenuItem:new( "ASelectItem4", "item4", "Ctrl+A", "LogoAtomic" ) )

 	mgmis = mis

end

function HandleUimenuitemEvent( eventType,  eventData)
    local refid = eventData["RefID"]:GetString()
    local widget = eventData["Target"]:GetPtr("AWidget")
    if widget == nil then
 		return
	end
    if eventData["Type"]:GetInt() == UI_EVENT_TYPE_CLICK then
        if widget:GetId() == "uimenuitemcode" then
            AppLog( "AMenuItem support : " .. widget:GetId() .. " was pressed " )
            ViewCode ( "LuaScripts/code_amenuitem.lua", widget:GetParent())
        end
        if widget:GetId() ==  "uimenuitemlayout" then
            AppLog( "AMenuItem support : " .. widget:GetId() .. " was pressed ")
            ViewCode ( "Scenes/layout_amenuitem.ui.txt", widget:GetParent())
        end

        if widget:GetId() ==  "uimenuitempush" then
			if mgmis ~= nil then
            	local mymenuwindow = AMenuWindow:new( widget, "MenuItemDemo")
           	 	SubscribeToEvent(mymenuwindow, "WidgetEvent", "HandleUimenuitemEvent" )
            	local xx = widget:GetX() + (widget:GetWidth()/2)
            	local yy = widget:GetY() + (widget:GetHeight()/2)
           	 	mymenuwindow:ShowMenu(mgmis, xx, yy)
			end
        end

        if widget:GetId() ==  "MenuItemDemo" then
            AppLog( "AMenuItem event : " .. widget:GetId() .. " and " .. refid .. " was selected ")
        end

        if widget:GetId() ==  "uimi1" then
            AppLog( "AMenuItem action : " .. widget:GetId() .. " was pressed ")
			if mgmis ~= nil then
            	mgmis:AddItem( AMenuItem:new( "New AMenuItem", "newitem") )
			end
        end
        if widget:GetId() ==  "uimi2" then
            AppLog( "AMenuItem action : " .. widget:GetId() .. " was pressed ")
			if mgmis ~= nil then
            	mgmis:AddItem( AMenuItem:new( "Newer AMenuItem", "neweritem" ) )
			end
        end
        if widget:GetId() ==  "uimi3" then
            AppLog( "AMenuItem action : " .. widget:GetId() .. " was pressed ")
			if mgmis ~= nil then
            	mgmis:AddItem( AMenuItem:new( "A Duck", "aduck", "", "DuckButton" ) )
			end
        end
        if widget:GetId() ==  "uimi4" then
            AppLog( "AMenuItem action : " .. widget:GetId() .. " was pressed ")
			if mgmis ~= nil then
            	mgmis:AddItem( AMenuItem:new( "Atomic!", "atomic", "", "LogoAtomic") )
			end
        end
        if widget:GetId() ==  "uimi5" then
            AppLog( "AMenuItem action : " .. widget:GetId() .. " was pressed ")
			if mgmis ~= nil then
           		mgmis:Clear()
			end
        end
    end
end

